### PR TITLE
technitium-dns-server: undo split / re-merge library and server

### PR DESCRIPTION
**Pull request by FabianRig**
## Things done

Update Technitium DNS Server to 13.3. More important: undo split / re-merge the library into the file so that automatic updates are possible. Refer to the closed PR at https://github.com/NixOS/nixpkgs/pull/363529 for details.

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/te/technitium-dns-server/package.nix [hunk 1]
```diff
@@ -29,7 +50,7 @@ buildDotnetModule rec {
   # move dependencies from TechnitiumLibrary to the expected directory
   preBuild = ''
     mkdir -p ../TechnitiumLibrary/bin
-    cp -r ${technitium-dns-server-library}/lib/${technitium-dns-server-library.pname}/* ../TechnitiumLibrary/bin/
+    cp -r ${technitium-library}/lib/TechnitiumLibrary/* ../TechnitiumLibrary/bin/
```
Comment (timhae) [inline pkgs/by-name/te/technitium-dns-server/package.nix:53]:
```suggestion
    cp -r ${technitium-library}/lib/TechnitiumLibrary/* ../${technitium-library.pname}/bin/
```

**Review by timhae — APPROVED** [review]

File: pkgs/by-name/te/technitium-dns-server/package.nix (hunk 1 above)
Comment (FabianRig) [reply pkgs/by-name/te/technitium-dns-server/package.nix:53]: Why do you think this might be helpful? I assume that the library will not be renamed by the original author?!

Comment (timhae) [reply pkgs/by-name/te/technitium-dns-server/package.nix:53]: the path will be whatever you have as pname for the package in the let binding, this will not change if you keep them in sync (you probably wouldn't change this ever, realisticly) but in the spirit of keeping it DRY I would reference that here instead of writing it again. the target path in the server program needs the library path to match whatever the builddotnetmodule sets it to and that will set it to the pname. so this does not depend on what the original author named it. you can try this out by changing the pname but not the reference here.

Comment (FabianRig) [reply pkgs/by-name/te/technitium-dns-server/package.nix:53]: OK, I think I got what you mean. However, I just created that directory with the line above it. 
```
mkdir -p ../TechnitiumLibrary/bin
```

So in case we want to replace it, we'd need to replace it in both lines. Correct? I see your argument. However, it makes the code a little less easy to read:

```
  preBuild = ''
    mkdir -p ../${technitium-library.pname}/bin
    cp -r ${technitium-library}/lib/TechnitiumLibrary/* ../${technitium-library.pname}/bin/
  '';
```

Comment (timhae) [reply pkgs/by-name/te/technitium-dns-server/package.nix]: Yes, correct. In the end it is up to you, I did that because when I extracted the library into its own package I had to dig down the layers of abstraction in builddotnetmodule (or more realistically: I opened a nix develop shell and looked around) because the build failed and that would be avoided by this change, hence my suggestion.

Comment (FabianRig) [reply pkgs/by-name/te/technitium-dns-server/package.nix]: I updated the PR accordingly -- in both places. :-)

Comment (timhae) [reply pkgs/by-name/te/technitium-dns-server/package.nix]: I am sorry for the confusion, I think the source dir needs to be set to the pname and you are absolutely right, the target dir which you create is whatever the upstream package name is. Sorry, I got confused by the diff..

Comment (timhae) [reply pkgs/by-name/te/technitium-dns-server/package.nix:53]: I replaced the wrong part in my suggestion, the source `TechnitiumLibrary` part should have been replaced by the library pname, sorry I'm on mobile at the moment and that males it really hard to parse the diffs

**Comment by FabianRig** [comment]
Yes... Because of https://github.com/NixOS/nixpkgs/commit/1cd9516969553291792021f1b327f0e2e2775b3c. -.-

I'll try to fix that.

**Comment by FabianRig** [comment]
> This now has a merge conflict.

Should be fixed now. Both nuget-deps are now jsons.

**Comment by FabianRig** [comment]
> This would allow us to test that the update of both together works well with the update script.

Thank you for your suggestion! However, the update script has never worked. Initially, the script's author suggested splitting the package -- which we did and which led to problems. Now we (re-)merge it in order for a new update script to work properly. 

So it would be great if you could merge it as is so that we can work on the new update script which updates both components simultaneously in one package.

**Comment by FabianRig** [comment]
> So, you split the package to make the update script work. It didn't work.
No. Another contributor asked to split the package and separate the library for an update script. I approved that PR. However, that attempt failed.


> Now you merge the package again - to make update script work. But we can't confirm.
No. This update script will probably not work -- it has never worked before. However, in the current state, the update script will never work. 

> This seems pointless. We should make sure that the update script works after merging the packages - before merging the PR.
It hasn't worked. Perhaps a new update script will work, but I don't know. I haven't created any update scripts yet.

> In any case, the commits should be split into merge and update - independently of whether we can confirm the update script working this way or not. Different kind of changes shouldn't be in the same commit.
OK. I removed the update part and only included the merging of the library into the package -- e.g. the original state.

**Comment by FabianRig** [comment]
Merge conflict removed.

**Comment by FabianRig** [comment]
> In short: I don't see any benefit to undo the split. It won't help with updating.

And I don't see any benefit in keeping an unnecessary split. I shouldn't have accepted the suggestion to split the package -- but unfortunately I did. I'd like to correct that now.

**Comment by GameDungeon** [comment]
Does this just need reviewers or is something else preventing this from moving forward?

**Comment by GameDungeon** [comment]
Having the two packages split is pointless. The library package is useless without this one, so it just adds an extra package cluttering up the nixpkgs repo. It just increases the maintenance burden by splitting the files up the files into different locations (possibly messing with updates too as both packages now need to be updated at the same time), for literally zero benefit. Generally it isn't a good idea to stand between maintainers and the packages they maintain without a good reason either.

**Comment by TroubleDog54** [comment]
Having two separate packages seems to be causing the DNS server to drag behind upstream. The package maintainer is fixing a mistake, can this please be merged?

Also to note: the possibility of having two PRs for two separate packages (as outlined in https://github.com/NixOS/nixpkgs/pull/364119#issuecomment-2541377885) just for updating one program seems rather messy. Having this PR merged will fix that potential issue.

**Review by TroubleDog54 — APPROVED** [review]

**Review by cyplo — APPROVED** [review]

**Comment by wolfgangwalther** [comment]
Here is an example of how to enable the update script for both together when they are merged into the same file:

https://github.com/NixOS/nixpkgs/blob/fd3bd33d9b4aec41a3b3696007fd58c815389395/pkgs/by-name/xe/xed/package.nix#L78-L86

If you set up the update script like this, I'm more than happy to merge the merge.

Edit: Or maybe not, because... this example doesn't even work. It seems to have been merged without testing it. See https://github.com/NixOS/nixpkgs/pull/365428#issuecomment-2585310599 - the `subpackage` option doesn't work in these cases. Sorry for the noise.
